File: src/proxtv.cc
/* File: proxtv.cc -*- c++ -*- */
/* (c) Copyright 2013  Alvaro Barbero and Suvrit Sra */
/* C++ library header */

#include "proxtv.h"

/* Solves: 0.5*norm(x-y)^2 + lam*TVL1(x)
 * 
 * INPUT
 * y --   vector of doubles
 * n --   length of y
 * lam -- regularization parameter
 *
 * OUTPUT
 * x -- solution vector
 * info -- status info about solver / solution
 *
 * ASSUMPTION
 * x -- preallocated array of length 'n'
 * 
 */
int tv1d (double* y, double* x, size_t n, double lam, double p, double* info)
{
  if (p==1) {
    TV1D_denoise(y, x, n, lam);
    info=0;
    return 0;
  } else if (p==2) { 
    return morePG_TV2(y, lam, x, info, n, NULL);
  } else {
    return GPFW_TVp(y, lam, x, info, n, p, NULL);
  }
}


/* Solves weighted-1D-TV
   0.5*norm(x-y)^2 + sum_i lam(i)*|x(i+1)-x(i)|
 * 
 * INPUT
 * y --   vector of doubles
 * n --   length of y
 * lam --   vector of weights parameter
 *
 * OUTPUT
 * x -- solution vector
 * info -- status info about solver / solution
 *
 * ASSUMPTION
 * x -- preallocated array of length 'n'
 * 
 */
int tv1dw(double* y, double* x, size_t n, double* lam, double* info)
{
  return PN_TV1_Weighted(y, lam, x, info, n, SIGMA, NULL);
}

/* Solves 2D-TV
 * 0.5*norm(x-y)^2 + sum_i lam*|x(i+1,j)-x(i,j)| + sum_j lam*|x(i,j+1)-x(i,j)|
 * 
 * INPUT
 * y --   vector of doubles
 * n --   length of y
 * lam -- regularization parameter
 * nthr --num of threads to use
 * maxit--max num of iterations of underlying solver
 *
 * OUTPUT
 * x -- solution vector
 * info -- status info about solver / solution
 *
 * ASSUMPTION
 * x -- preallocated array of length 'n'
 * 
 */
int tv2d (double* y, double* x, size_t m, size_t n, double* lam, double pr,
          double pc, int nthr, int maxit, double* info)
{
  return 0;
}

int tv2dw(double* y, size_t m, size_t n, double* lam, double pr,
          double pc, int nthr, int maxit, double* info)
{
  return 0;
}

File: src/proxtv.py
#
# Python module for using proxTV library functions
#
# (c) Copyright 2014  Alvaro Barbero and Suvrit Sra
#

import proxtv_internal as tv
import numpy as np

#
# [x,info] = TV(y,lambda,p,threads)  Proximity operator for Lp Total Variation.
#
# Solves the proximity problem associated with the n-dimensional Total Variation Lp norm.
# Depending on the dimension and norm of choice, a different algorithm is used for the
# optimization.
# Any norm p>=1 is accepted.
#
# Inputs:
#   - y: numpy multi-array, input of the proximity operator.
#   - lambda: premultiplier of the norm.
#   - [p]: norm (default 1).
#   - [threads]: number of threads (default 1). Used only for 2-D or larger signals.
#   - [mit]: max num iters for proximal Dykstra (default 10).
#
# Outputs:
#   - x: numpy multi-array, solution of the proximity problem.
#   - info: statistical info of the run algorithm (if any)
#       info.iters: number of iterations run (major iterations for the 2D case)
#       info.stop: value of the stopping criterion.
def TV(X, lambda, p, nthreads, mit)
    # Check input is numpy array
    if type(X) != np.ndarray:
        raise ProxTVInputArgumentException('X','Input signal must be numpy.ndarray object')
    
    # Check dimension of input signal
    dim = X.ndim
    
    #TODO
        
#
# Classes for proxTV exceptions
#

class ProxTVException(Exception):
    """Base class for exceptions in proxTV"""
    pass
    
class ProxTVInputArgumentException(ProxTVException):
    """Exception raised for errors the inputs when invoking proxTV functions.
    
    Attributes:
        expr -- input expression in which the error occurred
        msg -- explanation of the error
    """
    
    def __init__(self, expr, msg)
        self.expr = expr
        self.msg = msg

